compiler/rustc_mir_transform/src/coroutine/by_move_body.rs

```diff
@@ -302,20 +302,18 @@ impl<'tcx> MutVisitor<'tcx> for MakeByMoveBody<'tcx> {
                 projection
             };
 
-            // The only thing that should be left is a deref, if the parent captured
-            // an upvar by-ref.
-            std::assert_matches::assert_matches!(final_deref, [] | [mir::ProjectionElem::Deref]);
```

Comment:
This assertion only makes sense when we assume that captures are going to be *as precise as they can be*. This is not true when we truncate captures when they don't give us anything (`truncate_capture_for_optimization`), and also I'm pretty sure that edition 2018 doesn't even work like that at all.